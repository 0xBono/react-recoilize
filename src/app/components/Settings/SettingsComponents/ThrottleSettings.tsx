import React, {useState} from 'react';

const ThrottleSettings: React.FC = () => {
  // variable to store/reference connection
  let num = '';
  const onChange = (event: string) => {
    num = event;
  };
  // Creating function to tie to get to the backend
  const throttleFunc = () => {
    const backgroundConnection = chrome.runtime.connect();
    // post the message with index in payload to the connection
    backgroundConnection.postMessage({
      action: 'throttleEdit',
      tabId: chrome.devtools.inspectedWindow.tabId,
      payload: {value: num}, // edit this value to some other number
    });
  };
  return (
    <div>
      <h2>Enter Throttle</h2>
      <input onChange={e => onChange(e.target.value)} /> milliseconds
      <div>
        <button
          onClick={() => {
            throttleFunc();
          }}>
          Enter
        </button>
      </div>
    </div>
  );
};

export default ThrottleSettings;
