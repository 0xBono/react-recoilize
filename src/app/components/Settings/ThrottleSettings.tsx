import React, {useState, useEffect} from 'react';
import {useAppSelector, useAppDispatch} from '../../state-management/hooks';
import {
  newThrottle,
  resetThrottle,
} from '../../state-management/slices/ThrottleSlice';

const ThrottleSettings: React.FC = () => {
  const dispatch = useAppDispatch();

  const throttle = useAppSelector(state => state.throttle.throttleValue);
  console.log('Throttle', throttle);
  const [throttleNum, setThrottleNum] = useState<string>(throttle);
  console.log('Throttle Num', throttleNum);

  useEffect(() => {
    setThrottleNum(throttle);
  }, [throttle]);

  // onClick function for reset button. 70ms is the default throttle
  const onClick = (): void => {
    dispatch(resetThrottle());
  };

  // Creating function to tie to get to the backend
  const throttleFunc = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    const backgroundConnection = chrome.runtime.connect();
    // post the message with index in payload to the connection
    backgroundConnection.postMessage({
      action: 'throttleEdit',
      tabId: chrome.devtools.inspectedWindow.tabId,
      payload: {value: throttleNum}, // edit this value to some other number
    });

    dispatch(newThrottle(throttleNum));
  };

  return (
    <div>
      <h2>Enter Throttle</h2>
      <form onSubmit={throttleFunc}>
        <input
          type="text"
          style={{marginBottom: '10px'}}
          onChange={e => setThrottleNum(e.target.value)}
          value={throttleNum}
          placeholder="enter in milliseconds"
        />{' '}
        <span style={{fontSize: '14px'}}>milliseconds</span>
        <div>
          <button type="submit">Enter</button>
          <button onClick={onClick} type="button">
            Reset
          </button>
        </div>
      </form>
      <span>
        Current throttle is{' '}
        {throttle === '70' ? `${throttle} (default)` : throttle} ms
      </span>
    </div>
  );
};

export default ThrottleSettings;
