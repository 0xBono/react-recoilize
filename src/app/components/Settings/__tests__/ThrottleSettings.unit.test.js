import React, {useState} from 'react';
import ReactDOM from 'react-dom';
import {getQueriesForElement, getByText} from '@testing-library/dom';

import {render, fireEvent} from '@testing-library/react';

// importing necessary
import ThrottleSettings from '../ThrottleSettings';

// First cheeck if render without crashing
it('renders without crashing', () => {
  const root = document.createElement('div');
  ReactDOM.render(<ThrottleSettings />, root);
});

// todo: snapshot testing
it('renders & matches snapshot', () => {
  const {asFragment} = render(<ThrottleSettings />);
  expect(asFragment()).toMatchSnapshot();
});

it('renders to the dom with the correct content better', () => {
  const root = document.createElement('div');
  ReactDOM.render(<ThrottleSettings throttleDisplay={1000} />, root);
  const {getByText} = getQueriesForElement(root);
  expect(getByText('Enter Throttle')).not.toBeNull();
  expect(getByText('Enter')).not.toBeNull();
  expect(getByText('Reset')).not.toBeNull();
  expect(getByText('Current throttle is 1000 ms')).not.toBeNull();
});

it('Check the user input typing', () => {
  const {getByLabelText, getByText, debug} = render(
    <ThrottleSettings throttleDisplay={1000} />,
  );
  const input = getByLabelText('Enter Throttle');

  fireEvent.change(input, {target: {value: 500}}); // this successfully changes the value
  expect(input.value).toEqual('500');
});

// todo: Make sure the onclicks are working properly using jest mack
it('Check the onclick buttons', () => {
  // creating mock functions for chrome
  chrome.runtime.connect = function () {
    return {postMessage: function () {}};
  };
  let setThrottleDisplay = jest.fn();
  chrome.devtools = {inspectedWindow: {}};

  // Rendering the component
  const {getByText, debug} = render(
    <ThrottleSettings
      throttleDisplay={1000}
      setThrottleDisplay={setThrottleDisplay}
    />,
  );

  // Testing the buttons
  const submitButton = getByText('Enter'); // check these buttons -- how to test
  const resetButton = getByText('Reset');
  fireEvent.click(submitButton);
  fireEvent.click(resetButton);
});

// get the button to run mock function instead of actual
