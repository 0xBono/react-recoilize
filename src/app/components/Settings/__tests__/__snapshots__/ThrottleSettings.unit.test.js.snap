// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders & matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label>
        <h2>
          Enter Throttle
        </h2>
        <input
          style="margin-bottom: 10px;"
          type="text"
          value=""
        />
         
      </label>
      <span
        style="font-size: 14px;"
      >
        milliseconds
      </span>
      <div>
        <button
          type="submit"
        >
          Enter
        </button>
        <button
          type="submit"
        >
          Reset
        </button>
      </div>
    </form>
    <span>
      Current throttle is  ms
    </span>
  </div>
</DocumentFragment>
`;
